EchoPro/survey.py

```diff
@@ -327,269 +328,502 @@ def read_validated_data( self ,
+            .apply(lambda x: np.average( x['sigma_bs'] , weights=x['length_count']))
+            .to_frame('mean_sigma_bs')
+            .reset_index()
+        )
+                
+        # Now these values can be re-merged with stratum information and averaged over strata
+        mean_strata_sigma_bs = (
+            mean_haul_sigma_bs
+            .groupby(['stratum_num' , 'species_id'])['mean_sigma_bs']
+            .mean()
+            .reset_index()
+        )
+        
+        # Add back into object
+        self.acoustics['sigma_bs'] = {
+            'length_binned': all_length_df ,
+            'haul_mean': mean_haul_sigma_bs ,
+            'strata_mean': mean_strata_sigma_bs
+        }
+        
+        # Fill in missing sigma_bs values
+        self.impute_missing_sigma_bs( species_id )
     
+    def impute_missing_sigma_bs( self ,
+                                 species_id: np.float64 ):
+        """
+        Imputes sigma_bs for strata without measurements or values
+
+        Parameters
+        ----------
+        species_id : np.float64
+            Numeric code representing a particular species of interest
+
+        Notes
+        -----
+        This function iterates through all stratum layers to impute either the
+        nearest neighbor or mean sigma_bs for strata that are missing values.
+        """    
+        #### TODO: CURRENTLY : species_id is unused since only hake are being processed, but this will need
+        ### to actually be parameterized in the future
+        # Collect all possible strata values
+        strata_options = np.unique(self.spatial['strata_df'].copy().stratum_num)
+        
+        #
+        strata_mean = self.acoustics['sigma_bs']['strata_mean'].copy()
+        
+        # impute missing strata values
+        present_strata = np.unique(strata_mean['stratum_num']).astype(int)
+        missing_strata = strata_options[~(np.isin(strata_options, present_strata))]
+        
+        if len(missing_strata) > 0:            
+            sigma_bs_impute = pd.concat([strata_mean , 
+                                            pd.DataFrame({'stratum_num': missing_strata ,
+                                                         'species_id': np.unique( strata_mean.species_id ) ,
+                                                         'mean_sigma_bs': np.nan})]).sort_values('stratum_num')        
+            
+            # Find strata intervals to impute over        
+            for i in missing_strata:
+                strata_floor = present_strata[present_strata < i]
+                strata_ceil = present_strata[present_strata > i]
+
+                new_stratum_below = np.max(strata_floor) if strata_floor.size > 0 else None
+                new_stratum_above = np.min(strata_ceil) if strata_ceil.size > 0 else None      
+                
+                sigma_bs_indexed = sigma_bs_impute[sigma_bs_impute['stratum_num'].isin([new_stratum_below, new_stratum_above])]
```

Comment:
Not sure if I get this right, but seems that if `new_stratum_below` and `new_stratum_above` are not off by more than 1, the statement `.isin([new_stratum_below, new_stratum_above])]` would have some error.

Say if `new_stratum_below=2` and `new_stratum_above=5`, `.isin` would not get everything between 2 and 5.

Reply:
This calls for a separate cases in the same test function to make sure the imputation is correct, something like:
Case 1: new_stratum_below=2, new_stratum_above=4
Case 2: new_stratum_below=2, new_stratum_above=5

Reply:
For reference, this is what `self.acoustics['sigma_bs']['strata_mean']` should look like when no strata are missing: 

```python
   stratum_num  species_id  mean_sigma_bs
0          0.0       22500       0.000425
1          1.0       22500       0.000084
2          2.0       22500       0.000070
3          3.0       22500       0.000114
4          4.0       22500       0.000191
5          5.0       22500       0.000295
6          6.0       22500       0.000247
7          7.0       22500       0.000384
8          8.0       22500       0.000481
```

**Case 1**

Dropping `stratum_num = 3` via:

```python
strata_mean_test = strata_mean.loc[ strata_mean[ 'stratum_num' ] != 3 ]
``` 

yields for `sigma_bs_impute`:

```python
   stratum_num  species_id  mean_sigma_bs
0          0.0       22500       0.000425
1          1.0       22500       0.000084
2          2.0       22500       0.000070
0          3.0       22500            NaN
4          4.0       22500       0.000191
5          5.0       22500       0.000295
6          6.0       22500       0.000247
7          7.0       22500       0.000384
8          8.0       22500       0.000481
```

When running the loop nested within `for i in missing_strata: ...`, this then yields: 

```python
   stratum_num  species_id  mean_sigma_bs
0          0.0       22500       0.000425
1          1.0       22500       0.000084
2          2.0       22500       0.000070
0          3.0       22500       0.000130
4          4.0       22500       0.000191
5          5.0       22500       0.000295
6          6.0       22500       0.000247
7          7.0       22500       0.000384
8          8.0       22500       0.000481
```

which produces the expected result (i.e. the mean of strata 2 and 5). 

**Case 2**

Dropping `stratum_num = [ 3 , 4]` via:

```python
strata_mean_test = strata_mean.loc[ ~ strata_mean[ 'stratum_num' ].isin( [ 3 , 4 ] ) ]
```

does trigger an error when generating `sigma_bs_impute` due to the mismatch in the number of elements for each column. So this has been modified in the updated commit to correctly produce the following dataframe: 

```python
   stratum_num  species_id  mean_sigma_bs
0          0.0       22500       0.000425
1          1.0       22500       0.000084
2          2.0       22500       0.000070
0          3.0       22500            NaN
1          4.0       22500            NaN
5          5.0       22500       0.000295
6          6.0       22500       0.000247
7          7.0       22500       0.000384
8          8.0       22500       0.000481
```

When run through the rest of the function, this then yields the expected output where the values for strata 3 and 4 are filled with the averaged `mean_sigma_bs` from strata 2 and 5: 

```python
   stratum_num  species_id  mean_sigma_bs
0          0.0       22500       0.000425
1          1.0       22500       0.000084
2          2.0       22500       0.000070
0          3.0       22500       0.000183
1          4.0       22500       0.000183
5          5.0       22500       0.000295
6          6.0       22500       0.000247
7          7.0       22500       0.000384
8          8.0       22500       0.000481
```

**Additional cases**

When testing for the edge cases where the extreme strata values are missing via:

```python
strata_mean_test = strata_mean.loc[ ~ strata_mean[ 'stratum_num' ].isin( [ 0 , 8 ] ) ]
```
```python
0          0.0       22500            NaN
1          1.0       22500       0.000084
2          2.0       22500       0.000070
3          3.0       22500       0.000114
4          4.0       22500       0.000191
5          5.0       22500       0.000295
6          6.0       22500       0.000247
7          7.0       22500       0.000384
1          8.0       22500            NaN
```

The function then correctly imputes the nearest-neighbor values:

```python
   stratum_num  species_id  mean_sigma_bs
0          0.0       22500       0.000084
1          1.0       22500       0.000084
2          2.0       22500       0.000070
3          3.0       22500       0.000114
4          4.0       22500       0.000191
5          5.0       22500       0.000295
6          6.0       22500       0.000247
7          7.0       22500       0.000384
1          8.0       22500       0.000384
```

In a messier case where the edge cases are missing, a single stratum is missing from the dataset, and then back-to-back strata are also missing from the dataset:

```python
   stratum_num  species_id  mean_sigma_bs
0          0.0       22500            NaN
1          1.0       22500       0.000084
2          2.0       22500       0.000070
1          3.0       22500            NaN
4          4.0       22500       0.000191
2          5.0       22500            NaN
3          6.0       22500            NaN
7          7.0       22500       0.000384
4          8.0       22500            NaN
```

The function appears to impute the mean/nearest-neighbor `mean_sigma_bs` for the missing values: 

```python
   stratum_num  species_id  mean_sigma_bs
0          0.0       22500       0.000084
1          1.0       22500       0.000084
2          2.0       22500       0.000070
1          3.0       22500       0.000130
4          4.0       22500       0.000191
2          5.0       22500       0.000287
3          6.0       22500       0.000287
7          7.0       22500       0.000384
4          8.0       22500       0.000384
```

So this otherwise appears to function as intended when fixing the concatenation of the imputation dataframe.